types/react-query/index.d.ts

```diff
@@ -1,142 +1,370 @@
+    config?: QueryOptions<TResult>,
+): PaginatedQueryResult<TResult>;
+
+export function usePaginatedQuery<TResult, TKey extends AnyQueryKey, TVariables extends AnyVariables = []>({
+    queryKey,
+    variables,
+    queryFn,
+    config,
+}: {
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined);
+    variables?: TVariables;
+    queryFn: QueryFunctionWithVariables<TResult, TKey, TVariables>;
+    config?: QueryOptions<TResult>;
+}): PaginatedQueryResult<TResult>;
+
+// useInfiniteQuery
+export function useInfiniteQuery<TResult, TKey extends AnyQueryKey, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    queryFn: InfiniteQueryFunction<TResult, TKey, TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
 
-export type QueryFunction<TResult, TVariables extends object> = (variables: TVariables) => Promise<TResult>;
+export function useInfiniteQuery<TResult, TKey extends string, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    queryFn: InfiniteQueryFunction<TResult, [TKey], TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
 
-export interface QueryOptions<TResult> {
+export function useInfiniteQuery<TResult, TKey extends AnyQueryKey, TVariables extends AnyVariables, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    variables: TVariables,
+    queryFn: InfiniteQueryFunctionWithVariables<TResult, TKey, TVariables, TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
+
+export function useInfiniteQuery<TResult, TKey extends string, TVariables extends AnyVariables, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    variables: TVariables,
+    queryFn: InfiniteQueryFunctionWithVariables<TResult, [TKey], TVariables, TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
+
+export function useInfiniteQuery<
+    TResult,
+    TKey extends AnyQueryKey,
+    TMoreVariable,
+    TVariables extends AnyVariables = []
+>({
+    queryKey,
+    variables,
+    queryFn,
+    config,
+}: {
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined);
+    variables?: TVariables;
+    queryFn: InfiniteQueryFunctionWithVariables<TResult, TKey, TVariables, TMoreVariable>;
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>;
+}): InfiniteQueryResult<TResult, TMoreVariable>;
+
+export type QueryKeyPart = string | object | boolean | number | null | readonly QueryKeyPart[] | null | undefined;
+export type AnyQueryKey =
+    | readonly [QueryKeyPart]
```

Comment:
The first element of query key must always be `string` . With this type definition this code does not raise type error:

```
useQuery([10], (key) => {...} )
```

Reply:
I know that is written in the documentation but the implementation is really agnostic to what the key must be. It is just any serializable array. Moreover, the documentation is confusing and sometimes says contradicting things. For example, it says that the key must be `[string, object]` but have many examples where it deviates from that, such as examples where more elements or any other constituents allowed such as nested arrays. I concluded that the doc wasn't updated much while API was being changed and that requirement was from earlier versions.

Reply:
I actually agree that we can require the first element to be a string to make it clear that is the recommended way to go.
If you think it is better, I will make a change. Let me know.

Reply:
I think we should follow the [API docs](https://github.com/tannerlinsley/react-query#usequery) and not the internal implementation.

Reply:
ok, fair enough. I changed.